src/bci_build/package/__init__.py

```diff
@@ -1031,26 +1031,50 @@ def readme_url(self) -> str:
 
     @property
     def readme(self) -> str:
-        if "README.md" in self.extra_files:
-            if isinstance(self.extra_files["README.md"], bytes):
-                return self.extra_files["README.md"].decode("utf-8")
-            return str(self.extra_files["README.md"])
+        """The contents of the image README that is refered by ``io.artifacthub.package.readme-url`` label.
+
+        The text content is obtained from :py:attr:`BaseContainerImage.extra_files` if
+        ``README.md.j2`` is defined.
+
+        If no file ``README.md.j2`` is explicitly defined in :py:attr:`BaseContainerImage.extra_files`
+        but the file is available in the image folder, assuming :py:attr:`BaseContainerImage.name`
+        as the folder name, this file going to be used.
+
+        If no proper README is found, a default text based on :py:attr:`BaseContainerImage.title`
+        and :py:attr:`BaseContainerImage.description` will be returned.
+
+        """
+        if "README.md.j2" in self.extra_files:
+            tmpl = self.extra_files["README.md.j2"]
+
+            if isinstance(tmpl, bytes):
+                tmpl = tmpl.decode("utf-8")
+
+            return self.render_template(tmpl)
 
         readme_template_fname = Path(__file__).parent / self.name / "README.md.j2"
+
         if readme_template_fname.exists():
-            jinja2_env = jinja2.Environment(
-                loader=jinja2.FileSystemLoader(Path(__file__).parent / "templates"),
-                autoescape=jinja2.select_autoescape(["md"]),
-            )
-            return jinja2_env.from_string(readme_template_fname.read_text()).render(
-                image=self
-            )
+            return self.render_template(readme_template_fname.read_text())
 
         return f"""# The {self.title} Container image
 
 {self.description}
 """
 
+    def render_template(self, text):
```

Comment:
not a huge fan of another public method on the base class given that its only used in one place? also #1075  will change this

Reply:
I think I can skip this change.